agentstack/generation/tool_generation.py

```diff
@@ -139,3 +145,201 @@ def assert_tool_exists(tool_name: str, tools: dict):
+
+    return False
+
+
+def _process_tools_list(
+        current_tools: List[ast.AST],
+        tool_data: Dict,
+        operation: str,
+        base_name: str = 'tools'
+) -> List[ast.AST]:
+    """
+    Process a tools list according to the specified operation.
+
+    Args:
+        current_tools: Current list of tool nodes
+        tool_data: Tool configuration dictionary
+        operation: Operation to perform ('add' or 'remove')
+        base_name: Base module name for tools
+    """
+    if operation == 'add':
+        new_tools = current_tools.copy()
+        # Add new tools with bundling if specified
+        new_tools.extend(_create_tool_nodes(
+            tool_data["tools"],
+            tool_data.get("tools_bundled", False),
+            base_name
+        ))
+        return new_tools
+
+    elif operation == 'remove':
+        # Filter out tools that match any in the removal list
+        return [
+            tool for tool in current_tools
+            if not any(_is_tool_node_match(tool, name, base_name)
+                       for name in tool_data["tools"])
+        ]
+
+    raise ValueError(f"Unsupported operation: {operation}")
+
+
+def _modify_agent_tools(
+        node: ast.FunctionDef,
+        tool_data: Dict,
+        operation: str,
+        agents: Optional[List[str]] = None,
+        base_name: str = 'tools'
+) -> ast.FunctionDef:
+    """
+    Modify the tools list in an agent definition.
+
+    Args:
+        node: AST node of the function to modify
+        tool_data: Tool configuration dictionary
+        operation: Operation to perform ('add' or 'remove')
+        agents: Optional list of agent names to modify
+        base_name: Base module name for tools
+    """
+    # Skip if not in specified agents list
+    if agents is not None and node.name not in agents:
+        return node
+
+    # Check if this is an agent-decorated function
+    if not any(isinstance(d, ast.Name) and d.id == 'agent'
+               for d in node.decorator_list):
+        return node
+
+    # Find the Return statement and modify tools
+    for item in node.body:
+        if isinstance(item, ast.Return):
+            agent_call = item.value
+            if isinstance(agent_call, ast.Call):
+                for kw in agent_call.keywords:
+                    if kw.arg == 'tools':
+                        if isinstance(kw.value, ast.List):
+                            # Process the tools list
+                            new_tools = _process_tools_list(
+                                kw.value.elts,
+                                tool_data,
+                                operation,
+                                base_name
+                            )
+
+                            # Replace with new list
+                            kw.value = ast.List(elts=new_tools, ctx=ast.Load())
+
+    return node
+
+
+def modify_agent_tools(
+        framework: str,
+        tool_data: Dict,
+        operation: str,
```

Comment:
I see why you're passing the `operation` from the top, since a lot of the logic is shared in searching the tree to get to the point where you're making an addition or deletion.

Reply:
its annoying but it i had to go back and add that in so many layers because it matters at the very end